src/DriverFactory.php

```diff
@@ -123,4 +129,32 @@ private function registerAnnotationLoader(): void
 
         self::$isAnnotationLoaderRegistered = true;
     }
+
+    /**
+     * @param array<string, mixed> $config
+     */
+    private function createCachedReader(ContainerInterface $container, array $config, Reader $reader): PsrCachedReader
+    {
+        $cache = $this->retrieveDependency(
+            $container,
+            $config['cache'],
+            'cache',
+            CacheFactory::class
+        );
+
+        if ($cache instanceof Cache) {
+            $cache = CacheAdapter::wrap($cache);
+        }
+
+        if ($cache instanceof CacheItemPoolInterface) {
+            return new PsrCachedReader($reader, $cache);
+        }
+
+        $cacheType = is_object($cache) ? get_class($cache) : gettype($cache);
```

Comment:
I suggest moving this check into the exception named constructor to reduce the conditionals/complexity in here